lib/resty/worker/events.lua

```diff
@@ -406,30 +407,48 @@ _M.poll = function()
   return _M.poll()
 end
 
--- executes a polling loop, and reschedules the polling timer
+-- executes a polling loop
 local do_timer
 do_timer = function(premature)
-  if premature then
-    _M.post(_M.events._source, _M.events.stopping)
-  end
-
-  local ok, err = _M.poll()
-  if not ok then
-    log(ERR, "worker-events: timer-poll returned: ", err)
-  end
+  while true do
+    if premature then
+      _M.post(_M.events._source, _M.events.stopping)
+    end
 
-  if _interval ~= 0 and not premature then
-    ok, err = new_timer(_interval, do_timer)
+    local ok, err = _M.poll()
     if not ok then
-      if err == "process exiting" then
-        _M.post(_M.events._source, _M.events.stopping)
+      log(ERR, "worker-events: timer-poll returned: ", err)
+    end
+
+    if _interval ~= 0 and not premature then
+      if _interval <= 1 then
+        sleep(_interval)
+        premature = exiting()
+
+      else
+        local sleep_left = _interval
+        while sleep_left > 0 do
+          if sleep_left <= 1 then
+            sleep(sleep_left)
+            premature = exiting()
+            break
+          end
+
+          sleep(1)
+
+          premature = exiting()
+          if premature then
+            break
+          end
+
+          sleep_left = sleep_left - 1
+        end
```

Comment:
```suggestion
      local sleep_left = _interval
      while sleep_left > 0 do
        local sleep_step = min(sleep_left, 1)
        sleep (sleep_step)
        sleep_left = sleep_left - sleep_step

        premature = exiting()
        if premature then
          break
        end
      end
```
I always find this kind of code hard to follow (have to mentally trace the whole execution).  what about something like this?  IMO, it reads like a description: "while there's some sleep to do, sleep for at most one sec, account for it, and check if exiting at every step"

Reply:
it can be even simpler I think;
```lua
      local sleep_left = _interval
      while sleep_left > 0 do
        sleep (min(sleep_left, 1))
        sleep_left = sleep_left - 1
        premature = exiting()
        if premature then
          break
        end
      end
```
No need for the variable, `sleep_left` will go negative, and then fail `sleep_left > 0` and hence exit